extension/sigv4authextension/signingroundtripper.go

```diff
@@ -0,0 +1,118 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sigv4authextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/sigv4authextension"
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	sigv4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+	"go.uber.org/zap"
+)
+
+// SigningRoundTripper is a custom RoundTripper that performs AWS Sigv4.
+type SigningRoundTripper struct {
+	transport  http.RoundTripper
+	signer     *sigv4.Signer
+	region     string
+	service    string
+	creds      *aws.Credentials
+	awsSDKInfo string
+	logger     *zap.Logger
+}
+
+// RoundTrip() executes a single HTTP transaction and returns an HTTP response, signing
+// the request with Sigv4.
+func (si *SigningRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqBody, err := req.GetBody()
+	if err != nil {
+		return nil, err
+	}
+
+	content, err := ioutil.ReadAll(reqBody)
+	reqBody.Close()
+	if err != nil {
+		return nil, err
+	}
+	body := bytes.NewReader(content)
+
+	// Clone request to ensure thread safety.
+	req2 := cloneRequest(req)
+
+	// Add the runtime information to the User-Agent header of the request
+	ua := req2.Header.Get("User-Agent")
+	if len(ua) > 0 {
+		ua = ua + " " + si.awsSDKInfo
+	} else {
+		ua = si.awsSDKInfo
+	}
+	req2.Header.Set("User-Agent", ua)
+
+	// Sign the request
```

Comment:
```suggestion
	// hash the request
```

This creates a hash, not a signature or HMAC.  That comes later.